numpy/core/src/umath/loops.c.src

```diff
@@ -845,6 +846,109 @@ NPY_NO_EXPORT NPY_GCC_OPT_3 void
 
 /**end repeat1**/
 
+/*
+ * Arithmetic bit shift operations.
+ *
+ * Intel hardware masks bit shift values, so large shifts wrap around
+ * and can produce surprising results. The special handling ensures that
+ * behavior is independent of compiler or hardware.
+ * TODO: We could implement consistent behavior for negative shifts,
+ *       which is undefined in C.
+ */
+
+#define LEFT_SHIFT_OP                               \
+    do {                                            \
+        if (NPY_LIKELY(in2 < sizeof(@type@) * 8)) { \
+            *out = in1 << in2;                      \
+        }                                           \
+        else {                                      \
+            *out = 0;                               \
+        }                                           \
+    } while (0)
+
+
+NPY_NO_EXPORT NPY_GCC_OPT_3 void
+@TYPE@_left_shift(char **args, npy_intp *dimensions, npy_intp *steps,
+                  void *NPY_UNUSED(func))
+{
+    if (IS_BINARY_REDUCE) {
+        BINARY_REDUCE_LOOP(@type@) {
+            @type@ ip2_val = *(@type@ *)ip2;
```

Comment:
Could remove the reduce specialization, I suspect reduce using the shift operators would be rare and `BINARY_LOOP_FAST` should have adequate performance.

Reply:
Done in #13739